Big Data And Society: Lab 2
=====


# Forking the Repository

First, let's fork the course repo. A fork is a copy of a repository. Unlike cloning, when you fork a repo, it is attached to your account, and you are free to modify it, and edit, without affecting the original version... which is desirable for the context of the class.

1. First click fork on the github page
2. On GitHub, navigate to your fork of repository.
3. In the right sidebar of your fork's repository page, click  to copy the clone URL for your fork.
4. Open Terminal (for Mac and Linux users) or the command prompt (for Windows users).
5. Type git clone, and then paste the URL you copied in Step 2. 

### Updating your Version to get Materials from the Main Class Repository

To get the new materials from the main class repository, you need to commit any changes you've made locally and push them to your personal fork of the repo online. Then pull the repository from the main class repo.

1. Type git remote add upstream, and then paste the URL https://github.com/civic-data-design-lab/16_11.S947.git (you might get a message saying that this already exists, this is fine, it means this has already been done.)
2. To verify the new upstream repository you've specified for your fork, type **git remote -v**. You should see the URL for your fork as origin (remember your fork has **your username** in it), and the URL for the original repository as upstream (the one with the **civic-data-design-lab** in the URL).
3. Before you pull the upstream master repo, commit any changes you've made locally by following these steps:

===

* Stage commits using **git add .**
* Commit your staged changes using **git commit -m "my changes"**
* Push your work to your forked repo. **git push origin master** (if you get a warning about LF conflicting with CRLF, this is not a problem, you can proceed)

===
    
This will update your forked repo with any local changes.
    
4. Once this is done, **pull** the upstream repository (the main class repository). This will both fetch and merge the original repo into your local version. Do this using **git pull upstream master**

Pulling from the remote repository will bring in its branches and their respective commits.

# Introduction to Pandas
![IPython](images/pandas.png "Pandas")
**Pandas** is a library for Python for data manipulation and analysis. Pandas expands the data processing capacities of Python and adds a number of classes for easily importing data, in particular numerical tables, from various formats into their DataFrame object. A `DataFrame` is Panda’s basic object that allows multidimensional data processing and indexing. `DataFrames` can be easily and efficiently queried without the need of cumbersome syntax and convoluted loops. `DataFrames` can be merged with other data, they can be sliced, and they can be reshaped; in a way, we can think of Pandas as a big data combination of Excel and SQL. 

## Resources

Pandas employs a number of functional and declarative programming idioms making it a bit different from regulay Python syntax. However, it is very close to functionality to **NumPy**, treating objects as vectors. 

You can get a Cheatsheet to ease the learning curve, and learn about some of its functionality:

https://drive.google.com/folderview?id=0ByIrJAE4KMTtaGhRcXkxNHhmY2M&usp=sharing




### 0. Importing the library
First we need to import the library, and use a shorter name for it:

In [1]:
import pandas as pd
import numpy as np

### 1. Reading Files
Pandas provides a number of reader functions that generally return a pandas object or `DataFrame`. Multiplle different file types can be read, like **csv**, **txt**, **xls**, and **json**. The function parses the tabular data contained in the files and returns a formatted and indexed `DataFrame`. A number of additional arguments can be speficied, allowing to define the type of separator for text files, the header, the column to use as an index, etc. 

Additional documentation can be found on the API: 
http://pandas.pydata.org/pandas-docs/version/0.15.2/io.html


In [2]:
# We are reading a CSV with some Census data
df=pd.read_csv("data/2013arriyadh_pop_clean.csv")

# We can print the first 5 rows of the df
df.head()

TAZ  Saudi Male  Saudi Female  Total_Saudi  Non-Saudi Male  \
0    1         NaN           NaN          NaN             NaN   
1    2        5956          6943        12899            1943   
2    3        4339          4385         8724            2425   
3    4        1696          1160         2856            2427   
4    5        1710          1553         3263             620   

   Non-Saudi Female  Total_Non_Saudi  Total Pop  
0               NaN              NaN        NaN  
1              2738             4681      17580  
2              1189             3614      12338  
3               106             2533       5389  
4               914             1534       4797

Notice we have a table! A spreadsheet! And it indexed the rows. Pandas (borrowing from R) calls it a DataFrame. Lets see the types of the columns.

`df`, in Python context, is an instance of the `pd.DataFrame` class, created by calling the `pd.read_csv` function, which calls the DataFrame constructor inside of it. `df` is a dataframe object.



### 2. DataFrame methods:
The df object, for example, has methods, or functions belonging to it, which allow it to do things. For example `df.head()` is a method that shows the first 5 rows of the dataframe. `df.dtypes` returns the data types of each of the columns. 

Additional documentation can be found here: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html
In general, the API is the best resource: http://pandas.pydata.org/pandas-docs/stable/api.html

In [3]:
df.dtypes

TAZ                 float64
Saudi Male          float64
Saudi Female        float64
Total_Saudi         float64
Non-Saudi Male      float64
Non-Saudi Female    float64
Total_Non_Saudi     float64
Total Pop           float64
dtype: object

The shape of the object is:

In [4]:
df.shape

(1497, 8)

1497 rows times 8 columns. A spredsheet is a matrix. `df.shape` returns a tuple, so we can access members of this **tuple** like we do with a list:

In [5]:
df.shape[0], df.shape[1]

(1497, 8)

These are the column names:

In [6]:
df.columns

Index([u'TAZ', u'Saudi Male', u'Saudi Female', u'Total_Saudi', u'Non-Saudi Male', u'Non-Saudi Female', u'Total_Non_Saudi', u'Total Pop'], dtype='object')

Notice that `df.columns` returns a `Pandas.Series` object. This object is built on top of the Python **lists**, and has similar methods and attributes. 

### 3. Querying

Pandas `DataFrames` have built-in methods for performing queries in a SQL style. They can be diced, sorted, etc it. We can apply any of this queries to parts of the `df`, and based on this query select subsets of the whole `df`. Here we look for all the TAZ with a Total Populations of less than 300.

In [7]:
df['Total Pop'] < 300

0     False
1     False
2     False
3     False
4     False
5     False
6      True
7     False
8     False
9     False
10     True
11    False
12    False
13    False
14    False
...
1482     True
1483    False
1484    False
1485    False
1486    False
1487    False
1488     True
1489    False
1490     True
1491     True
1492    False
1493    False
1494    False
1495    False
1496    False
Name: Total Pop, Length: 1497, dtype: bool

This gives us a `Pandas.Series` of `True`s and `False`s. We call this a mask. If we count the number of `True`s with a population of less than 300, we'll get the fraction of TAZ $\lt$ 300. 

To implement this, we can combine the Pandas query with a **Numpy** function:

In [8]:
np.sum(df['Total Pop'] < 300)

357

Notice that you could just find the average since the `True`s equal to 1 and `False`s to 0.


In [9]:
np.mean(df['Total Pop'] < 300)

0.23847695390781562

Or directly, in Pandas, which works since `df['Total Pop'] < 300` is a pandas Series. Pandas series have a number of built-in methods like `.mean()`!

In [10]:
(df['Total Pop'] < 300).mean()

0.23847695390781562

#### Exercise:
Calculate the minimum value, and maximum value of the `Total Pop` column. [This](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.min.html#pandas.Series.min) documentation might be useful.

In [1]:
# Your code here


### 4. Filtering

We can use the `df` mask to get a filtered dataframe: we use the mask to "index" into the dataframe to get the rows we want.

In [11]:
df[df['Total Pop'] < 300]

TAZ  Saudi Male  Saudi Female  Total_Saudi  Non-Saudi Male  \
6        7          33            29           62              77   
10      11           0             0            0               0   
22      23          35            32           67              19   
25      26           0             0            0               0   
38      39           2             7            9              60   
40      41           3             8           11              81   
47      48          56            45          101              46   
49      50         195            97          292               3   
50      51         109            89          198               2   
59      60           1             1            2               2   
67      68           0             0            0               0   
69      70           2             2            4              11   
74      75           0             0            0               0   
96      97           7             5           12              25   
98      99          25            24           49             107   
110    111           2             1            3               7   
112    113           8             7           15              34   
114    115           3             3            6              42   
116    117           1             0            1               3   
149    150           7             7           14               7   
150    151           0             0            0               0   
160    161           0             0            0               0   
183    184           0             0            0               7   
185    186          10             7           17             132   
214    215          31            34           65             140   
221    222           3             3            6              17   
224    225           0             0            0               0   
225    226          18            12           30              67   
227    228           2             2            4              15   
239    240           0             0            0               0   
...    ...         ...           ...          ...             ...   
1366  1367          36            32           68              55   
1396  1397           0             0            0               0   
1397  1398           0             0            0               0   
1401  1402           0             0            0               0   
1404  1405           4             3            7              15   
1405  1406           1             0            1               3   
1411  1412           8             7           15               0   
1412  1413           0             0            0               0   
1413  1414           0             0            0               0   
1416  1417          69            53          122               3   
1431  1432         101            80          181              26   
1434  1435           0             0            0              17   
1436  1437           2             0            2              62   
1440  1441           0             0            0               0   
1441  1442           0             0            0               0   
1442  1443           0             0            0               0   
1443  1444          23            22           45               1   
1444  1445           3             2            5               1   
1445  1446         118           112          230              19   
1447  1448          35            36           71              44   
1453  1454         129           117          246               2   
1455  1456          75            94          169              30   
1456  1457           0             0            0               0   
1465  1466           1             1            2              11   
1466  1467           2             1            3              19   
1472  1473           2             0            2             189   
1482 

We can query based on multiple conditions, by using a `&` condition. All we need to do is add `()` brackets around each condition. The query uses a boolean AND. Each condition ceates a mask of trues and falses.

In [12]:
df[(df['Total Pop'] < 300) & (df['Total_Non_Saudi'] > 70)]

TAZ  Saudi Male  Saudi Female  Total_Saudi  Non-Saudi Male  \
6        7          33            29           62              77   
38      39           2             7            9              60   
40      41           3             8           11              81   
47      48          56            45          101              46   
98      99          25            24           49             107   
185    186          10             7           17             132   
214    215          31            34           65             140   
225    226          18            12           30              67   
267    268           9             8           17              54   
270    271          11            10           21              63   
282    283           7             6           13             173   
306    307          27            22           49             103   
318    319           6             3            9              91   
397    398           5             7           12             129   
410    411           0             0            0              77   
434    435          38            36           74              17   
450    451           0             0            0              83   
451    452           1             0            1             189   
452    453           1             0            1              81   
496    497          35            36           71              49   
509    510           0             0            0             264   
549    550           0             0            0              81   
556    557           0             0            0             175   
564    565           0             0            0             184   
610    611           7             0            7             249   
617    618          45            41           86              53   
784    785          96            85          181              52   
852    853          48            38           86              77   
854    855          25            20           45              88   
931    932          88            77          165              42   
1029  1030           9             7           16              53   
1041  1042          58            57          115              40   
1051  1052          13            14           27              67   
1053  1054          21            22           43             108   
1060  1061           3             2            5              84   
1065  1066           6             4           10              80   
1066  1067          23            18           41              74   
1095  1096          73            73          146              84   
1097  1098          27            19           46              70   
1131  1132          46            49           95              53   
1133  1134          12            13           25              81   
1135  1136           6             6           12             103   
1142  1143          79            48          127              45   
1156  1157          37            25           62             105   
1160  1161          23            15           38              93   
1161  1162          58            39           97              54   
1176  1177          26            18           44             134   
1198  1199          76            69          145              60   
1201  1202          15            14           29             158   
1205  1206          19            18           37             144   
1219  1220           5             4            9             146   
1220  1221           4             3            7             116   
1288  1289          65            43          108             122   
1290  1291          61            40          101             127   
1320  1321          57            49          106             111   
1361  1362          36            32           68             135   
1447  1448          35            36           71              44   
1472 

#### Exercise:
From the total population, subtract the total non-saudi male population, and get the average across the whole dataset. [Documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.subtract.html#pandas.Series.subtract)

In [2]:
# Your code here


### 5. Cleaning

One of the most common tasks while working with big data is data cleaning. Datasets will be inherently heterogeneous and unstructured. The lack of structure can cause functions to throw errors. We can clean data sets through various methods; if the discrepancies follow a structured pattern, it is possible to use built-in functions, or write our own functions. However, if the errors cannot be structured, we might have to do it manually!

For this example, we first check the datatypes:


In [13]:
df.dtypes

TAZ                 float64
Saudi Male          float64
Saudi Female        float64
Total_Saudi         float64
Non-Saudi Male      float64
Non-Saudi Female    float64
Total_Non_Saudi     float64
Total Pop           float64
dtype: object

A common cause of errors is having Null or 'None' values. This usually means data was missing. 

In [14]:
df[df['Total Pop'].isnull()]

TAZ  Saudi Male  Saudi Female  Total_Saudi  Non-Saudi Male  \
0       1         NaN           NaN          NaN             NaN   
1492  NaN         NaN           NaN          NaN             NaN   
1493  NaN         NaN           NaN          NaN             NaN   
1494  NaN         NaN           NaN          NaN             NaN   
1495  NaN         NaN           NaN          NaN             NaN   
1496  NaN         NaN           NaN          NaN             NaN   

      Non-Saudi Female  Total_Non_Saudi  Total Pop  
0                  NaN              NaN        NaN  
1492               NaN              NaN        NaN  
1493               NaN              NaN        NaN  
1494               NaN              NaN        NaN  
1495               NaN              NaN        NaN  
1496               NaN              NaN        NaN

We had some incomplete data! We can get rid of it with the `.notnull()` function.

In [15]:
df = df[df['Total Pop'].notnull()]
df.shape

(1491, 8)

We removed those 6 rows. 

### 6. Split-apply-combine
With Pandas we can split the data into groups based on some criteria. We can then apply a function to each group independently, and finally combine the results into a new data structure.

<img src=http://i.imgur.com/yjNkiwL.png></img>

#### Find Road Network Nodes and Edges
Let’s look at a network file of Riyadh. The network was generated based on cell phone data, and it represents cell phone towers in the area, and the flow among them. In this example, we are going to:

* Split data per node ID. We will be grouping the dataset by their source node ID.
* sum their voc (volume/capacity). Once we had grouped them, we can sum by source node.
* combine the result. We can finally perform operations on the grouped dataset.


In [16]:
# Import the file
edges=pd.read_csv("data/riyadh_edges.csv")
edges.head()

# Group the edges by their source
grouped_sources = edges['gid'].groupby(edges['source'])

## count and combine
edges_per_node = grouped_sources.count()
edges_per_node

source
1         3
2         3
3         1
4         4
5         1
6         1
7         1
8         1
9         2
10        2
11        3
12        3
13        2
14        1
15        2
...
10714     1
10715     1
10716     1
10717     2
10718     1
10719     1
10720     1
10721     1
10722     1
10723     1
10724     1
10725     1
10726     1
10727     1
10728     1
Name: gid, Length: 10726, dtype: int64

In [17]:
## average the flow per source and combine
grouped_voc_source = edges['voc'].groupby(edges['source'])
average_voc_source = grouped_voc_source.mean()
print "Average VOC Source Nodes:"
print average_voc_source
print

# Get the maximum source node
maximum_voc_source = average_voc_source.max()
slow_voc_source = average_voc_source[average_voc_source == maximum_voc_source].index

Average VOC Source Nodes:
source
1         0.002690
2         0.000585
3         0.028723
4         0.028389
5         0.161425
6         0.188712
7         0.051845
8         0.188947
9         0.000532
10        0.008032
11        0.000526
12        0.000000
13        0.003684
14        0.006842
15        0.004720
...
10714     0.047895
10715     0.000431
10716     0.011228
10717     0.000000
10718     0.003770
10719     0.114737
10720     0.032281
10721     0.000000
10722     0.010000
10723     0.000000
10724     0.031053
10725     0.016842
10726     0.003860
10727     0.037197
10728     0.027368
Name: voc, Length: 10726, dtype: float64



In [18]:
print 'The slowest edge is: ', 
print slow_voc_source


# what are we doing with slow source node?
print "Slow Source Node"
print edges[edges.gid.isin(slow_voc_source)].source
print

The slowest edge is:  Int64Index([10576], dtype='int64')
Slow Source Node
202    3755
Name: source, dtype: int64

